community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/InvocationTracer.java

```diff
@@ -0,0 +1,282 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.util.dbstructure;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.neo4j.helpers.Pair;
+import org.neo4j.kernel.api.constraints.UniquenessConstraint;
+import org.neo4j.kernel.api.index.IndexDescriptor;
+
+import static java.lang.String.format;
+
+public class InvocationTracer<C> implements InvocationHandler
+{
+    private static final String LINE_SEPARATOR = System.getProperty( "line.separator" );
+
+    private final String generatedClassPackage;
+    private final String generatedClassName;
+
+    private final StringBuilder builder;
+    private final Class<C> interfaceClass;
+
+    public InvocationTracer( String generatedClassNameWithPackage, Class<C> interfaceClass )
+    {
+        if ( generatedClassNameWithPackage.contains( "%" ) )
+        {
+            throw new IllegalArgumentException( "Format character in generated class name: " + generatedClassNameWithPackage );
+        }
+
+        Pair<String, String> parsedGenerated = parseQualifiedClassName( generatedClassNameWithPackage );
+        this.generatedClassPackage = parsedGenerated.first();
+        this.generatedClassName = parsedGenerated.other();
+
+        this.interfaceClass = interfaceClass;
+
+        this.builder = new StringBuilder();
+    }
+
+    public C newProxy()
+    {
+        return newProxy( interfaceClass );
+    }
+
+    public <P extends C> P newProxy( Class<P> proxyClass )
+    {
+        ClassLoader classLoader = proxyClass.getClassLoader();
+        return proxyClass.cast(
+            Proxy.newProxyInstance( classLoader, new Class[] { proxyClass }, this )
+        );
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder result = new StringBuilder();
+        printPreamble( result );
+        printTrace( result );
+        printAppendix( result );
+        return result.toString();
+    }
+
+    public String getGeneratedPackageName()
+    {
+        return generatedClassPackage;
+    }
+
+    public String getGeneratedClassName()
+    {
+        return generatedClassName;
+    }
+
+    private void printPreamble( StringBuilder builder )
+    {
+        String interfaceClassName = interfaceClass.getCanonicalName();
+        println( builder, "package %s;", generatedClassPackage );
+        println( builder );
+        println( builder, "import org.neo4j.kernel.api.constraints.UniquenessConstraint;");
+        println( builder, "import org.neo4j.kernel.api.index.IndexDescriptor;");
+        println( builder, "import org.neo4j.kernel.impl.util.dbstructure.Visitable;");
```

Comment:
How about using `Class.getName()` here to allow for refactoring? ([The comment in Visitable](https://github.com/neo4j/neo4j/pull/4023/files#diff-49ccc3a6cc436dc8953e3754ce02e59dR23) hints that this would be a good idea...)